crates/red_knot_python_semantic/src/semantic_index/ast_ids.rs

```diff
@@ -142,36 +294,50 @@ impl HasScopedExpressionId for ast::ExprRef<'_> {
 
 #[derive(Debug, Default)]
 pub(super) struct AstIdsBuilder {
-    expressions_map: FxHashMap<ExpressionNodeKey, ScopedExpressionId>,
-    uses_map: FxHashMap<ExpressionNodeKey, ScopedUseId>,
+    expressions: FxHashMap<ExpressionNodeKey, ScopedExpressionId>,
+    uses: FxHashMap<ExpressionNodeKey, ScopedUseId>,
+    eager_nested_scopes: FxHashMap<EagerNestedScopeNodeKey, ScopedEagerNestedScopeId>,
 }
 
 impl AstIdsBuilder {
     /// Adds `expr` to the expression ids map and returns its id.
     pub(super) fn record_expression(&mut self, expr: &ast::Expr) -> ScopedExpressionId {
-        let expression_id = self.expressions_map.len().into();
+        let expression_id = self.expressions.len().into();
 
-        self.expressions_map.insert(expr.into(), expression_id);
+        self.expressions.insert(expr.into(), expression_id);
 
         expression_id
     }
 
     /// Adds `expr` to the use ids map and returns its id.
     pub(super) fn record_use(&mut self, expr: &ast::Expr) -> ScopedUseId {
-        let use_id = self.uses_map.len().into();
+        let use_id = self.uses.len().into();
 
-        self.uses_map.insert(expr.into(), use_id);
+        self.uses.insert(expr.into(), use_id);
 
         use_id
     }
 
+    pub(super) fn record_eager_nested_scope(
+        &mut self,
+        node: impl Into<EagerNestedScopeNodeKey>,
+    ) -> ScopedEagerNestedScopeId {
+        let nested_scope_id = self.eager_nested_scopes.len().into();
+        *self
+            .eager_nested_scopes
+            .entry(node.into())
+            .or_insert(nested_scope_id)
```

Comment:
We use this pattern so that when building, we don't have to worry about whether we've already registered the eager nested scope in any particular containing scope